lib/SILOptimizer/Mandatory/SemanticARCOpts.cpp

```diff
@@ -665,6 +666,84 @@ bool SemanticARCOptVisitor::visitCopyValueInst(CopyValueInst *cvi) {
 //                         load [copy] Optimizations
 //===----------------------------------------------------------------------===//
 
+/// A simple analysis that checks if a specific def (in our case an inout
+/// argument) is ever written to. This is conservative, local and processes
+/// recursively through the def->use.
+static bool defUseIsAddressWrittenTo(SILFunctionArgument *arg) {
+  SmallVector<Operand *, 8> worklist(arg->getUses());
+  while (!worklist.empty()) {
+    auto *op = worklist.pop_back_val();
+    SILInstruction *user = op->getUser();
+
+    if (Projection::isAddressProjection(user)) {
+      for (SILValue r : user->getResults()) {
+        llvm::copy(r->getUses(), std::back_inserter(worklist));
+        continue;
+      }
+    }
+
+    if (auto *oeai = dyn_cast<OpenExistentialAddrInst>(user)) {
+      // Mutable access!
+      if (oeai->getAccessKind() != OpenedExistentialAccess::Immutable) {
+        return true;
+      }
+
+      //  Otherwise, look through it and continue.
+      llvm::copy(oeai->getUses(), std::back_inserter(worklist));
+      continue;
+    }
+
+    // load_borrow and debug_value_addr are both fine as well.
+    if (isa<LoadBorrowInst>(user) || isa<DebugValueAddrInst>(user)) {
+      continue;
+    }
+
+    // As long as we do not have a load [take], we are fine.
+    if (auto *li = dyn_cast<LoadInst>(user)) {
+      if (li->getOwnershipQualifier() == LoadOwnershipQualifier::Take) {
+        return true;
+      }
+      continue;
+    }
+
+    // If we have a FullApplySite, see if we use the value as an
+    // indirect_guaranteed parameter. If we use it as inout, we need
+    // interprocedural analysis that we do not perform here.
+    if (auto fas = FullApplySite::isa(user)) {
+      if (fas.getArgumentConvention(*op) ==
+          SILArgumentConvention::Indirect_In_Guaranteed)
```

Comment:
Too bad you need to hard-code `Full-Apply + Indirect_In_Guaranteed`. I feel like we should be using the `isMutatingOrConsuming` logic from the verifier to handle closures, coroutines, and other conventions... but it looks like it returns an overly optimistic results for `inout_aliasable`

Reply:
I was just trying to do something really simple. I am going to refactor it out.